### ISSUE TITLE
be clearer on mutability

### ISSUE THREAD

**Issue by mimoo** [zksecurity/noname#134]
we're not super clear on what can be mutated and what cannot be.

For example, can this function call mutate `thing`?

```rust
let mut thing = 5;
stuff(thing);
```

imo we should make mutability when we're passing mutable values around. In rust we would have to write `stuff(&mut thing)`. In noname everything is passed by value so I'm not sure what would be a good syntax (there is no such thing as `stuff(&thing)` for example. Maybe `stuff(mut thing)`?

How about this example:

```rust
fn stuff(arg: Field) { /* ... */ }
```

can the body of `stuff` mutate `arg`? I think it could only if we have `fn stuff(mut arg: Field)`, which I don't think is something we currently have in functions/methods.

The best thing to do would be to create tests to showcase these examples

**Comment by katat** [comment]
`fn stuff(mut arg: Field)` makes sense to mutate an external variable inside the function. 

atm, all the variables passing between the functions are done via copying their values. The reference variable is only usable at local scope.

**Comment by mimoo** [comment]
arg, so now I'm wondering what this does exactly if `stuff` mutates `a`:

```rust
let mut a = 5;
stuff(a);
```

**Comment by katat** [comment]
The function `stuff(a)` can be called with that mutable variable. Inside that function, that variable `a` is seen as immutable. As long as it doesn't directly mutate the argument `a` which is not mutable as the syntax `mut` is not supported for the function argument, it won't throw error.

**Comment by bufferhe4d** [comment]
I faced this problem when I was planning to test the example given in #224 . 

Below, even if the room variable is mut, we are not able to write functions like update_size since it is not possible to pass self as mut. I don't know if this is necessary to have but just found it relevant to share.

```
struct Room {
	pub beds: Field, // public
        size: Field // private
}

Room.update_size(self, size: Field) {
        self.size = size; 
        ^ error: self is not mutable
}

fn main() {
       let mut room = Room {beds: 2, size: 10};
       room.beds = 2; // allowed
       room.size = 5;  // not allowed
       room.update_size(5); // allowed
}
```

**Comment by mimoo** [comment]
I think we're in a model where functions/methods have to produce a new object instead of updating the current object. It's more "functional" and doesn't really save you any constraints underneath so maybe we should keep it this way?

**Comment by mimoo** [comment]
yeah I'm not against that path either, but the problem is... what keyword do we introduce for a mutable ref? `&mut` sounds ugly when we don't have references in the first place. Maybe `Room.update(mut self, ...)` would do?

**Comment by mimoo** [comment]
mmm, I would still vote for `refmut` because I personally would probably be confused by the `mut` alone

**Comment by katat** [comment]
Cool. Let's go with the `mutref` for now. cc @bufferhe4d

**Comment by bufferhe4d** [comment]
> Cool. Let's go with the `mutref` for now. cc @bufferhe4d

Alright, I am on it.
